common/planners/src/plan_read_datasource.rs

```diff
@@ -16,18 +16,45 @@ use std::collections::BTreeMap;
 use std::sync::Arc;
 
 use common_datavalues::DataField;
+use common_datavalues::DataSchema;
 use common_datavalues::DataSchemaRef;
 use common_meta_types::TableInfo;
 
 use crate::Expression;
 use crate::Extras;
 use crate::Partitions;
+use crate::S3ExternalTableInfo;
 use crate::Statistics;
 
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq)]
+pub enum SourceInfo {
+    // Normal table source, 'fuse/system'.
+    TableSource(TableInfo),
+
+    // S3 external source, 's3://'.
+    S3ExternalSource(S3ExternalTableInfo),
```

Comment:
Why not make `s3` as a special table? 

Such as a table function that holds the temporary table.

Reply:
Make s3 as table engine, we can do the unloading as `copy from table into s3://`

Reply:
It seems not extensible enough, considering we will support `azblob`, `gcs`, and other locations.

Reply:
But table function can also support insert.

`copy from table into s3://`

  `select * from s3://`

1. `as_table` to generate a temporary table.
2.  Select or insert works with the table.
3.

Reply:
I have add a desc for why make it as a table engine.

Reply:
BTW, for `select * from @stage ...`, we will convert the `@stage` alias to the external table info which get from the metasrv(after we support stage) in the sql parser phase, it will be same as `select * from s3://` did.